include/dice/template-library/ranges.hpp

```diff
@@ -2,40 +2,466 @@
-				std::forward<Pred>(pred));                                         \
+// all_of, any_of, none_of terminal algorithm
+#define DTL_DEFINE_PIPEABLE_RANGE_ALGO(ALGO_NAME)                                     \
+	namespace ranges_algo_detail {                                                    \
+		template<typename Pred>                                                       \
+		struct ALGO_NAME##_fn {                                                       \
+			Pred pred;                                                                \
+                                                                                      \
+			template<typename P>                                                      \
+			constexpr explicit ALGO_NAME##_fn(P &&p) : pred(std::forward<P>(p)) {}    \
+                                                                                      \
+			/* input_range causes no dangling risk since this adaptor is terminal  */ \
+			template<std::ranges::input_range R>                                      \
+			constexpr bool                                                            \
+			operator()(R &&r) const {                                                 \
+				return std::ranges::ALGO_NAME(r, pred);                               \
+			}                                                                         \
+                                                                                      \
+			template<typename R>                                                      \
+			friend auto operator|(R &&r, const ALGO_NAME##_fn &self)                  \
+					-> decltype(self(std::forward<R>(r))) {                           \
+				return self(std::forward<R>(r));                                      \
+			}                                                                         \
+		};                                                                            \
+	}                                                                                 \
+                                                                                      \
+	template<typename Pred>                                                           \
+	constexpr auto ALGO_NAME(Pred &&pred) {                                           \
+		return ranges_algo_detail::ALGO_NAME##_fn<std::decay_t<Pred>>(                \
+				std::forward<Pred>(pred));                                            \
 	}
 
-
 namespace dice::template_library {
 	DTL_DEFINE_PIPEABLE_RANGE_ALGO(all_of)
 	DTL_DEFINE_PIPEABLE_RANGE_ALGO(any_of)
 	DTL_DEFINE_PIPEABLE_RANGE_ALGO(none_of)
 }// namespace dice::template_library
+
+// empty / non_empty terminal algorithm
+namespace dice::template_library {
+	namespace ranges_algo_detail {
+		struct empty_fn {
+			template<std::ranges::input_range R>
+			constexpr bool operator()(R &&range) const {
+				// Use std::ranges::empty if the expression is valid for the range type.
+				if constexpr (requires { std::ranges::empty(range); }) {
```

Comment:
`std::ranges::empty` always works as far as I can tell. It has the expression in the `else` branch as a fallback

https://www.en.cppreference.com/w/cpp/ranges/empty.html

Reply:
see here https://github.com/dice-group/tentris-lib/pull/586/commits/396f6ce89c6b253ff30f866250d307f902fa0eec

in the left-join operator. 
```c++
if (std::ranges::empty(all_var_ids)){
```

didn't compile while:
```c++
if (std::ranges::begin(all_var_ids) == std::ranges::end(all_var_ids)){
```